ipa-core/src/protocol/ipa_prf/validation_protocol/proof_generation.rs

```diff
@@ -0,0 +1,517 @@
+    protocol::{
+        context::{
+            dzkp_field::{UVTupleBlock, BLOCK_SIZE},
+            Context,
+        },
+        ipa_prf::malicious_security::{
+            lagrange::{CanonicalLagrangeDenominator, LagrangeTable},
+            prover::{LargeProofGenerator, SmallProofGenerator, UVValues},
+        },
+        prss::SharedRandomness,
+        RecordId,
+    },
+    secret_sharing::SharedValue,
+};
+
+/// This is a tuple of `ZeroKnowledgeProofs` owned by a verifier.
+/// It consists of a proof from the prover on the left
+/// and a proof from the prover on the right.
+///
+/// The first proof has a different length, i.e. length `P`.
+/// It is therefore not stored in the vector with the other proofs.
+///
+/// `ProofsToVerify` also contains two masks `p_0` and `q_0` in `masks` stored as `(p_0,q_0)`
+/// These masks are used as additional `u,v` values for the final proof.
+/// These masks mask sensitive information when verifying the final proof.
+#[derive(Debug)]
+pub struct BatchToVerify {
+    first_proof_from_left: [Fp61BitPrime; LargeProofGenerator::PROOF_LENGTH],
+    first_proof_from_right: [Fp61BitPrime; LargeProofGenerator::PROOF_LENGTH],
+    proofs_from_left: Vec<[Fp61BitPrime; SmallProofGenerator::PROOF_LENGTH]>,
+    proofs_from_right: Vec<[Fp61BitPrime; SmallProofGenerator::PROOF_LENGTH]>,
+    masks: (Fp61BitPrime, Fp61BitPrime),
+}
+
+impl BatchToVerify {
+    /// This function generates the `BatchToVerify`.
+    /// The helper party fist generates its own proof
+    /// together with the proof from the left, i.e. `proof_from_left` (via `PRSS`)
+    /// It then distributes
+    /// shares of the proof to the helper on the left
+    /// and
+    /// receives a proof from the right helper, i.e. `proof_from_right`.
+    pub async fn generate_batch_to_verify<C, I>(ctx: C, uv_tuple_block: I) -> Self
+    where
+        C: Context,
+        I: Iterator<Item = UVTupleBlock<Fp61BitPrime>> + Clone,
+    {
+        // generate Proof
+        let (prover_left_proof, proof_from_left, (p, q)) =
+            ProofBatch::compute_proof(&ctx, uv_tuple_block);
+
+        // send prover_left_proof and receive proof_from_right
+        let length = prover_left_proof.len();
+        let send = prover_left_proof.send_to_left(&ctx);
+        let receive = ProofBatch::receive_from_right(&ctx, length);
+        let ((), proof_from_right) = try_join(send, receive).await.unwrap();
+
+        // output
+        BatchToVerify {
+            first_proof_from_left: proof_from_left.first_proof,
+            first_proof_from_right: proof_from_right.first_proof,
+            proofs_from_left: proof_from_left.proofs,
+            proofs_from_right: proof_from_right.proofs,
+            masks: (p, q),
+        }
+    }
+}
+
+/// This a `ProofBatch` generated by a prover.
+struct ProofBatch {
+    first_proof: [Fp61BitPrime; LargeProofGenerator::PROOF_LENGTH],
+    proofs: Vec<[Fp61BitPrime; SmallProofGenerator::PROOF_LENGTH]>,
+}
```

Comment:
Are we sure we want to hard-code the assumption that the first iteration is one size while all others are something else? This isn't very flexible.